payu/models/mom6.py

```diff
@@ -78,3 +80,31 @@ def init_config(self):
             input_nml['SIS_input_nml']['input_filename'] = input_type
 
         f90nml.write(input_nml, input_fpath, force=True)
+
+    def add_parameter_config_files(self):
+        """Check that the parameter files listed in input.nml are in the
+        model's configuration files"""
+        input_nml = f90nml.read(os.path.join(self.work_path, 'input.nml'))
+
+        # Set of all configuration files
+        config_files = set(self.config_files).union(self.optional_config_files)
+
+        for input in ['MOM_input_nml', 'SIS_input_nml']:
+            input_namelist = input_nml.get(input, {})
+            parameter_files = input_namelist.get('parameter_filename', [])
+
+            if isinstance(parameter_files, str):
+                parameter_files = [parameter_files]
+
+            for filename in parameter_files:
+                if filename not in config_files:
+                    print(f"payu: warning: parameter file {filename} "
+                          f"listed under {input} in input.nml is not in "
+                          "mom6's configuration files")
+
+                    # Extend config files
+                    self.config_files.append(filename)
+
+                    # Copy parameter file from control path to work path
+                    file_path = os.path.join(self.control_path, filename)
+                    shutil.copy(file_path, self.work_path)
```

Comment:
This makes me think we should fix up the (optional) config files definitions. For an ocean-only run, the only _required_ files are `input.nml` and `diag_table`. In fact, soon (but as far as I can tell, not yet), this could be either `diag_table` or `diag_table.yaml`, if using a YAML-enabled FMS.

I don't think we have the ability to distinguish between coupled ice-ocean and ocean-only runs, so `data_table` and `field_table` should be the only optional config files. I don't think the `data_table.xxx` files are ever read by the model, they were probably included because they exist in the [MOM6-examples](https://github.com/NOAA-GFDL/MOM6-examples) repository. Once again, with a YAML-enabled FMS, it might be the case that `data_table.yaml` could be valid.

The `MOM_override`, `MOM_layout`, `MOM_saltrestore`, `SIS_input`, `SIS_override` and `SIS_layout` entries could probably all be removed: they're now dynamically determined by this code. Instead of raising a warning if a file is specified in `input.nml` but _not_ in the (optional) config files, I think it should raise a warning if there aren't any entries in `MOM_input_nml` (or `SIS_input_nml`) at all. Perhaps in the latter case, it would be if the key is present but empty, as an ocean-only run wouldn't have that key present at all, and we probably don't want to be raising a warning in that case.

Reply:
Thanks for the review! I am looking into implementing the suggested changes now. Regarding diag_table/diag_table.yaml (and data_table/data_table.yaml), would there be any way to know what option is required other than just checking that there is one of them present in the control directory? Also, would I leave `MOM_input` out of the hard-coded `config_files` or is it useful enough for guaranteeing that `MOM_input` is always present?

Reply:
> Regarding diag_table/diag_table.yaml (and data_table/data_table.yaml), would there be any way to know what option is required other than just checking that there is one of them present in the control directory?

I think the YAML configs are very cutting-edge, I'd probably not worry about it for now. I don't know that there would be a good way to figure out what the model will read though, given it depends on whether FMS is compiled with/without a specific preprocessor define!

> Also, would I leave `MOM_input` out of the hard-coded `config_files` or is it useful enough for guaranteeing that `MOM_input` is always present?

It would be fairly convention-defying to use a different name for the main configuration file, but I don't see anything wrong with doing that. Since you're populating `self.config_files` at this point anyway, I guess there'll be a guarantee that _all_ the required parameter files are present, regardless of the default? I suppose that's why I suggested a warning if there aren't any parameter files in the namelist at all.

Reply:
Thanks, I've added a commit that tidies up the config_files and optional_config_files, and refactored it slightly so if the `diag_table/diag_table.yaml` logic could easily be added to the `add_config_files` method if needed. I've added in a warning if there aren't any parameter files for MOM_input_nml (and  SIS_input_nml if present).